Migrate topic log compaction (#402)
aiven/devportal: docs/products/kafka/howto/configure-log-cleaner.rst

Configure log cleaner for compaction
=====================================

By default the log cleaner is enabled, but the log compaction is disabled. To enable log compaction, follow these steps:

#. Log into the `Aiven web console <https://console.aiven.io/>`_ and select your Aiven for Apache Kafka service.
#. Scroll down to *Advanced configuration* and click on **Add configuration option**.
#. Find ``log.cleanup.policy`` in the list and select it.
#. Set the value to ``compact``.

> This change will affect all topics in the cluster that do not have a configuration override in place.

Reviewer: I believe this needs to be a warning, since it can affect existing topics and create disasters

> Control the frequency and delay

Reviewer: should the title state that is **log cleaning frequency and delay**?

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Before the cleaning begins, the cleaner thread will inspect the logs to find those with highest **dirty ratio** which is calculated as the ratio of the number of bytes in the head vs the total number of bytes in the log (tail + head). The default ratio value is 50%, but you can change it by change value of the property ``kafka.log_cleaner_min_cleanable_ratio`` from the *Advanced configuration* list.

Reviewer: Why is this ratio used? Adding a bit more explanation telling that this ratio allows to understand how many duplicated keys we have in a topic, and those will need to be compacted?

Configure log cleaner for topic compaction
==========================================

> The log cleaner is the process ensuring only the most recent value for a certain message key is kept within a partition for :doc:`compacted topics <../concepts/log-compaction>`.  In Aiven for Apache Kafka the log cleaner is enabled, but the log compaction is disabled by default. To enable log compaction, follow these steps:

Reviewer 2: ```suggestion
The log cleaner is the process ensuring only the most recent value for a certain message key is kept within a partition for :doc:`compacted topics <../concepts/log-compaction>`.  In Aiven for Apache Kafka, the **log cleaner** is enabled, but the **log compaction** is disabled by default. To enable log compaction, follow these steps:
```

> #. Log into the `Aiven web console <https://console.aiven.io/>`_ and select your Aiven for Apache Kafka service.

Reviewer 2: I like that those steps is consistent across the files. I mean how the format is. 💯

#. Scroll down to *Advanced configuration* and click on **Add configuration option**.
#. Find ``log.cleanup.policy`` in the list and select it.
#. Set the value to ``compact``.

.. warning:: This change will affect all topics in the cluster that do not have a configuration override in place.

Log cleaning frequency and delay
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Before the cleaning begins, the cleaner thread will inspect the logs to find those with highest **dirty ratio** calculated as the number of bytes in the head vs the total number of bytes in the log (tail + head); you can read more about head and tail definition in the :doc:`compacted topic documentation <../concepts/log-compaction>`. The ratio provides an estimate of how many duplicated keys are present in a topic, and therefore need to be compacted.

Reviewer 2: ```suggestion
Before the cleaning begins, the cleaner thread will inspect the logs to find those with highest **dirty ratio** calculated as the number of bytes in the head vs the total number of bytes in the log (tail + head); you can read more about head and tail definition in the :doc:`compacted topic documentation <../concepts/log-compaction>`. The ratio provides an estimation of how many duplicated keys are present in a topic; therefore, needs to be compacted.
```

> Before the cleaning begins, the cleaner thread will inspect the logs to find those with highest **dirty ratio** calculated as the number of bytes in the head vs the total number of bytes in the log (tail + head); you can read more about head and tail definition in the :doc:`compacted topic documentation <../concepts/log-compaction>`. The ratio provides an estimate of how many duplicated keys are present in a topic, and therefore need to be compacted.

Reviewer 2: thinking about one thing versus another, do you think would be better to have a shorter version and with bold to make it easy to find: for example, **kafka vs rabbitmq** but for this piece here: 

> the number of bytes in the head vs the total number of bytes in the log (tail + head)

I do not have a suggestion but maybe if you have it, feel free to shorter

.. Tip::

>     For the log cleaner to start compacting a topic, the dirty ratio needs to be bigger than a threshold set to 50% by default. You can change this value by modifying the property ``kafka.log_cleaner_min_cleanable_ratio`` in the *Advanced configuration* section of the service overview.

Reviewer 2: ```suggestion
    For the log cleaner to start compacting a topic, the dirty ratio needs to be bigger than the 50% default threshold. You can change this value by modifying the property ``kafka.log_cleaner_min_cleanable_ratio`` in the *Advanced configuration* section of the service overview.
```

> The log cleaner can be configured to leave some amount of not compacted "head" of the log by setting compaction time lag. You can achieve this by setting two additional properties from the  *Advanced configuration*:

Reviewer 2: ```suggestion
The log cleaner can be configured to leave some amount of not compacted "head" of the log by setting compaction time lag. You can achieve this by setting two additional properties from the *Advanced configuration*:
```

> -  ``log.cleaner.min.compaction.lag.ms`` : setting to a value greater than 0 will prevent log cleaner from compacting messages with an age newer than a minimum message age. Allows to delay compacting records.

Reviewer 2: does it make sense to connect those two sentences with an **and**? What does allows to delay compacting records: the process itself or the setting the value greater than 0?

> -  ``log.cleaner.max.compaction.lag.ms`` : the maximum amount of time message will remain not compacted. Please note, that exact time of compression can be later when comparing to the maximum compaction lag, depending on the availability log cleaner threads and actual compaction time.

Reviewer 2: ```suggestion
-  ``log.cleaner.max.compaction.lag.ms`` : the maximum amount of time a message will remain not compacted. Please note, that exact time of compression can be later when comparing to the maximum compaction lag, depending on the availability log cleaner threads and actual compaction time.
```
the availability **of** log cleaner threads? or is this a term itself?

Tombstone records
~~~~~~~~~~~~~~~~~

During the cleanup process, log cleaner threads also removes records that have a null value, also known as **tombstone** records. These records can be delayed from being deleted by configuring ``delete.retention.ms`` for the compacted topic.

> Consumers are able to read all tombstone messages as long as they reach the head of the topic before the period defined in ``delete.retention.ms`` (default 24 hours) is passed.

Reviewer 2: ```suggestion
Consumers are able to read all tombstone messages as long as they reach the head of the topic before the period defined in ``delete.retention.ms`` (default: 24 hours) is passed.
```

Configure log cleaner for topic compaction
==========================================

The log cleaner is the process ensuring only the most recent value for a certain message key is kept within a partition for :doc:`compacted topics <../concepts/log-compaction>`.  In Aiven for Apache Kafka the log cleaner is enabled, but the **log compaction** is disabled by default. To enable log compaction, follow these steps:

#. Log into the `Aiven web console <https://console.aiven.io/>`_ and select your Aiven for Apache Kafka service.
#. Scroll down to *Advanced configuration* and click on **Add configuration option**.
#. Find ``log.cleanup.policy`` in the list and select it.
#. Set the value to ``compact``.

> .. warning:: This change will affect all topics in the cluster that do not have a configuration override in place.

Reviewer 3: Can I configure some or all of these options per topic? I am not sure if that's what you mean, that this will only affect the topics that don't have specific configuration. Can we make it clearer which things can be configured at topic level and where to find the options for that?

Author: Rephrased and added more information on global and per-topic configuration
